Update NavigationObstacles for Godot 4.3 changes (#9336)
godotengine/godot-docs: tutorials/navigation/navigation_using_navigationobstacles.rst

Using NavigationObstacles
=========================

2D and 3D versions of NavigationObstacles nodes are available as
:ref:`NavigationObstacle2D<class_NavigationObstacle2D>` and
:ref:`NavigationObstacle3D<class_NavigationObstacle3D>`  respectively.

> Navigation obstacles are dual purpose in that they can affect both the navigation mesh baking and the agent avoidance.

Reviewer: ```suggestion
Navigation obstacles are dual purpose in that they can affect both the navigation mesh baking, and the agent avoidance.
```

.. figure:: img/nav_mesh_obstacles_discard.webp
   :align: center
   :alt: Navigation obstacles discard of unwanted navigation mesh

   Navigation obstacles discard of unwanted navigation mesh.

An obstacle does not add geometry in the baking process, it only removes geometry.
It does so by nullifying all the (voxel) cells with rasterized source geometry that are within the obstacle shape.
As such its effect and shape detail is limited to the cell resolution used by the baking process.

For more details on the navigation mesh baking see :ref:`doc_navigation_using_navigationmeshes`.

.. image:: img/nav_mesh_obstacles_properties.webp

The property ``affect_navigation_mesh`` makes the obstacle contribute to the navigation mesh baking.
It will be parsed or unparsed like all other node objects in a navigation mesh baking process.

The ``carve_navigation_mesh`` property makes the shape unaffected by offsets of the baking, 
e.g. the offset added by the navigation mesh ``agent_radius``.
It will basically act as a stencil and cut into the already offset navigation mesh surface.
It will still be affected by further postprocessing of the baking process like edge simplification.

> The obstacle shape and placement is defined with the ``height`` and ``vertices`` properties and the ``global_position`` of the obstacle.

Reviewer: ```suggestion
The obstacle shape and placement is defined with the ``height`` and ``vertices`` properties, and the ``global_position`` of the obstacle.
```

The y-axis value of any Vector3 used for the vertices is ignored as the obstacle is projected on a flat horizontal plane.

> For baking navigation mesh in scripts obstacles can be added procedural as a projected obstruction.

Reviewer: ```suggestion
When baking navigation meshes in scripts obstacles can be added procedurally as a projected obstruction.
```

- When used dynamically NavigationObstacles push away avoidance controlled agents in a radius around them.

Static avoidance obstacles
~~~~~~~~~~~~~~~~~~~~~~~~~~

An avoidance obstacle is considered static when its ``vertices`` property is populated with an outline array of positions to form a polygon.

.. figure:: img/nav_static_obstacle_build.gif
   :align: center
   :alt: Static obstacle drawn in the editor to block or contain navigation agents

   Static obstacle drawn in the editor to block or contain navigation agents.

- Static obstacles act as hard do-not-cross boundaries for avoidance using agents, e.g. similar to physics collision but for avoidance.
- Static obstacles define their boundaries with an array of outline ``vertices`` (positions), and in case of 3D with an additional ``height`` property.
- Static obstacles only work for agents that use the 2D avoidance mode.
- Static obstacles define through winding order of the vertices if agents are pushed out or sucked in.
- Static obstacles can not change their position. They can only be warped to a new position and rebuilt from scratch.

>   Static obstacles as a result are ill-suited for usages where the position is changed every frame as the constant rebuild has a high performance cost.

Reviewer: ```suggestion
  Static obstacles as a result are ill-suited for usages where the position is changed every frame, as the constant rebuild has a high performance cost.
```

Using NavigationObstacles
=========================

2D and 3D versions of NavigationObstacles nodes are available as
:ref:`NavigationObstacle2D<class_NavigationObstacle2D>` and
:ref:`NavigationObstacle3D<class_NavigationObstacle3D>`  respectively.

> Navigation obstacles are dual purpose in that they can affected both the navigation mesh baking and the agent avoidance.

Reviewer 2: ```suggestion
Navigation obstacles are dual purpose in that they can affect both the navigation mesh baking and the agent avoidance.
```

- When used statically NavigationObstacles constrain avoidance controlled agents outside or inside a polygon defined area.
- When used dynamically NavigationObstacles push away avoidance controlled agents in a radius around them.

Static avoidance obstacles
~~~~~~~~~~~~~~~~~~~~~~~~~~

An avoidance obstacle is considered static when its ``vertices`` property is populated with an outline array of positions to form a polygon.

.. figure:: img/nav_static_obstacle_build.gif
   :align: center
   :alt: Static obstacle drawn in the editor to block or contain navigation agents

   Static obstacle drawn in the editor to block or contain navigation agents.

- Static obstacles act as hard do-not-cross boundaries for avoidance using agents, e.g. similar to physics collision but for avoidance.
- Static obstacles define their boundaries with an array of outline ``vertices`` (positions), and in case of 3D with an additional ``height`` property.
- Static obstacles only work for agents that use the 2D avoidance mode.
- Static obstacles define through winding order of the vertices if agents are pushed out or sucked in.

> - Static obstacles can not change their position. They can only be warped to a new position and rebuild from scratch.

Reviewer 2: ```suggestion
- Static obstacles can not change their position. They can only be warped to a new position and rebuilt from scratch.
```

  Static obstacles as a result are ill-suited for usages where the position is changed every frame as the constant rebuild has a high performance cost.
- Static obstacles that are warped to another position can not be predicted by agents. This creates the risk of getting agents stuck should a static obstacle be warped on top of agents.

When the 2D avoidance is used in 3D the y-axis of Vector3 vertices is ignored. Instead, the global y-axis position of the obstacle is used as the elevation level. Agents will ignore static obstacles in 3D that are below or above them. This is automatically determined by global y-axis position of both obstacle and agent as the elevation level as well as their respective height properties.

Dynamic avoidance obstacles
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. figure:: img/obstacle_avoidance.gif
   :align: center
   :alt: Navigation agents deterred by a dynamic navigation obstacle's radius

>    Navigation agents deterred by a dynamic navigation obstacle's radius

Reviewer 2: ```suggestion
   Navigation agents deterred by a dynamic navigation obstacle's radius.
```

- Dynamic obstacles are not a reliable way to constrain agents in crowded or narrow spaces.

While both static and dynamic properties can be active at the same time on the same obstacle this is not recommended for performance.
Ideally when an obstacle is moving the static vertices are removed and instead the radius activated.
When the obstacle reaches the new final position it should gradually enlarge its radius to push all other agents away.
With enough created safe space around the obstacle it should add the static vertices again and remove the radius.

> This helps to avoid getting agents stuck in the suddenly appearing static obstacle when the rebuilt static boundary is finished.

Reviewer: ```suggestion
This helps avoid getting agents stuck in the suddenly appearing static obstacle when the rebuilt static boundary is finished.
```

Similar to agents the obstacles can make use of the ``avoidance_layers`` bitmask.
All agents with a matching bit on their own avoidance mask will avoid the obstacle.

Procedural obstacles
--------------------

> New obstacles can be created in script without a Node by using the NavigationServer directly.

Reviewer: ```suggestion
New obstacles can be created in a script without a Node by using the NavigationServer directly.
```

- Dynamic obstacles are not a reliable way to constrain agents in crowded or narrow spaces.

While both static and dynamic properties can be active at the same time on the same obstacle this is not recommended for performance.
Ideally when an obstacle is moving the static vertices are removed and instead the radius activated.
When the obstacle reaches the new final position it should gradually enlarge its radius to push all other agents away.

> With enough created save space around the obstacle it should add the static vertices again and remove the radius.

Reviewer 2: ```suggestion
With enough created safe space around the obstacle it should add the static vertices again and remove the radius.
```

> This helps to avoid getting agents stuck in the suddenly appearing static obstacle when the rebuild static boundary is finished.

Reviewer 2: ```suggestion
This helps to avoid getting agents stuck in the suddenly appearing static obstacle when the rebuilt static boundary is finished.
```
